src/arm/debugger/debugger.c

```diff
@@ -69,10 +107,42 @@ static bool ARMDebuggerUpdateStackTraceInternal(struct mDebuggerPlatform* d, uin
 		}
 		destAddress = info.op1.immediate + cpu->gprs[ARM_PC];
 	} else if (info.operandFormat & ARM_OPERAND_REGISTER_1) {
-		if (!isCall && info.op1.reg != ARM_LR && !(_ARMModeHasSPSR(cpu->cpsr.priv) && info.op1.reg == ARM_PC)) {
-			return false;
+		if (isCall) {
+			destAddress = cpu->gprs[info.op1.reg];
+		} else {
+			bool isExceptionReturn = _ARMModeHasSPSR(cpu->cpsr.priv) && info.affectsCPSR && info.op1.reg == ARM_PC;
+			bool isMovPcLr = (info.operandFormat & ARM_OPERAND_REGISTER_2) && info.op1.reg == ARM_PC && info.op2.reg == ARM_LR;
+			bool isBranch = ARMInstructionIsBranch(info.mnemonic);
+			int reg = (isBranch ? info.op1.reg : info.op2.reg);
+			destAddress = cpu->gprs[reg];
+			if (isBranch || (info.op1.reg == ARM_PC && !isMovPcLr)) {
+				// ARMv4 doesn't have the BLX opcode, so it uses an assignment to LR before a BX for that purpose.
+				struct ARMInstructionInfo prevInfo;
+				if (cpu->executionMode == MODE_ARM) {
+					ARMDecodeARM(cpu->memory.load32(cpu, pc - 4, NULL), &prevInfo);
+				} else {
+					ARMDecodeThumb(cpu->memory.load16(cpu, pc - 2, NULL), &prevInfo);
+				}
+				if ((prevInfo.operandFormat & ARM_OPERAND_REGISTER_1) && (prevInfo.operandFormat & ARM_OPERAND_AFFECTED_1) && prevInfo.op1.reg == ARM_LR) {
```

Comment:
`(prevInfo.operandFormat & (ARM_OPERAND_REGISTER_1 | ARM_OPERAND_AFFECTED_1))  == (ARM_OPERAND_REGISTER_1 | ARM_OPERAND_AFFECTED_1)` is fewer checks, and since this is a short circuit I'm not sure the compiler will optimize them together. (Though it may, since `prevInfo` isn't volatile.)

Reply:
Yeah, I couldn't really decide which form of repeating myself was less offensive. I'll switch it.